core/src/androidMain/kotlin/Peripheral.kt

```diff
@@ -0,0 +1,291 @@
+    public override val services: List<DiscoveredService>?
+        get() = platformServices?.map { it.toDiscoveredService() }
+
+    @Volatile
+    private var _connection: Connection? = null
+    private val connection: Connection
+        inline get() = _connection ?: throw NotReadyException(toString())
+
+    private val connectJob = atomic<Job?>(null)
+
+    private fun createConnectJob(): Job = scope.launch(start = LAZY) {
+        val connection = bluetoothDevice.connect(context, _state).also { _connection = it }
+        if (connection == null) {
+            _events.emit(Rejected)
+            return@launch
+        }
+
+        try {
+            connection
+                .characteristicChanges
+                .onEach(observers.characteristicChanges::emit)
+                .catch { cause -> if (cause !is ConnectionLostException) throw cause }
+                .onCompletion { _events.emit(Event.Disconnected) }
+                .launchIn(scope, start = UNDISPATCHED)
+
+            suspendUntilConnected()
+            discoverServices()
+            observers.rewire()
+            _events.emit(Event.Ready)
+        } catch (t: Throwable) {
+            connection.close()
+            _connection = null
+            if (t !is ConnectionLostException) throw t
+        }
+    }.apply {
+        invokeOnCompletion { connectJob.value = null }
+    }
+
+    public override suspend fun connect() {
+        check(!job.isCancelled) { "Cannot connect, scope is cancelled for $this" }
+        connectJob.updateAndGet { it ?: createConnectJob() }!!.join()
+    }
+
+    public override suspend fun disconnect() {
+        try {
+            job.cancelAndJoinChildren()
+            connection.bluetoothGatt.disconnect()
+            suspendUntilDisconnected()
+            connectJob.value = null
+        } finally {
+            _connection?.close()
+            _connection = null
+        }
+    }
+
+    public override suspend fun rssi(): Int = connection.request<OnReadRemoteRssi> {
+        readRemoteRssi()
+    }.rssi
+
+    private suspend fun discoverServices() {
+        connection.request<OnServicesDiscovered> {
+            discoverServices()
+        }
+        platformServices = connection.bluetoothGatt
+            .services
+            .map { it.toPlatformService() }
+            .also(::log)
+    }
+
+    private fun log(services: List<PlatformService>) {
+        services.forEach { service ->
+            Log.d(TAG, "• Service: ${service.serviceUuid}")
+            service.characteristics.forEach { characteristic ->
+                Log.d(TAG, "    • Characteristic: ${characteristic.characteristicUuid}")
+                characteristic.descriptors.forEach { descriptor ->
+                    Log.d(TAG, "        • Descriptor: ${descriptor.descriptorUuid}")
+                }
+            }
+        }
+    }
+
+    public suspend fun requestMtu(mtu: Int): Unit = connection.request {
+        requestMtu(mtu)
+    }
```

Comment:
This looks infinitely recursive, am I missing something?

Reply:
Oh, that is super confusing.

`connection.request {` makes `this` the underlying Android `BluetoothGatt`; so the `requestMtu` in the lambda is calling `BluetoothGatt`'s function, not the `Peripheral` function here.

I could change `connection.request` signature to have the parameter (rather than `this`) be the `BluetoothGatt`, but that would make everywhere we do this more verbose. Worth it?

Reply:
Could you make it `this@request.requestMtu`? That makes it clear at sight that it's not the same `requestMtu` function being called, without leaking into the other uses of request.

Reply:
Perfect, thanks for the suggestion. My train of thought was way too complicated. 🙃